src/eth/TaraClient.sol

```diff
@@ -135,8 +150,17 @@ contract TaraClient is IBridgeLightClient {
      */
     function processValidatorChanges(PillarBlock.VoteCountChange[] memory validatorChanges) public {
         for (uint256 i = 0; i < validatorChanges.length; i++) {
-            validators[validatorChanges[i].validator] += validatorChanges[i].change;
-            totalWeight += validatorChanges[i].change;
+            if (validatorChanges[i].change > 0) {
+                validatorVoteCounts[validatorChanges[i].validator] += uint256(uint32(validatorChanges[i].change));
+                totalWeight += uint256(uint32(validatorChanges[i].change));
+            } else {
+                if (validatorVoteCounts[validatorChanges[i].validator] < uint256(uint32(-validatorChanges[i].change))) {
+                    validatorVoteCounts[validatorChanges[i].validator] = 0;
+                } else {
+                    validatorVoteCounts[validatorChanges[i].validator] -= uint256(uint32(-validatorChanges[i].change));
+                }
+                totalWeight -= uint256(uint32(-validatorChanges[i].change));
+            }
```

Comment:
Maybe it is worth it to move to a separate function, something like `add(uint256 a, int32 b) uint256`